kubetest2-gke/deployer/firewall.go

```diff
@@ -27,53 +27,88 @@ import (
-		return fmt.Errorf("error creating e2e firewall: %v", err)
+	return ensureFirewallRulesForMultiProjects(d.projects, d.network, d.subnetworkRanges)
+}
+
+// Ensure firewall rules for e2e testing for all clusters in one single project.
+func ensureFirewallRulesForSingleProject(project, network string, clusters []string, instanceGroups map[string]map[string][]*ig) error {
+	for _, cluster := range clusters {
+		klog.V(1).Infof("Ensuring firewall rules for cluster %s in %s", cluster, project)
+		firewall := getClusterFirewall(project, cluster, instanceGroups)
+		if runWithNoOutput(exec.Command("gcloud", "compute", "firewall-rules", "describe", firewall,
+			"--project="+project,
+			"--format=value(name)")) == nil {
+			// Assume that if this unique firewall exists, it's good to go.
+			continue
+		}
+		klog.V(1).Infof("Couldn't describe firewall '%s', assuming it doesn't exist and creating it", firewall)
+
+		tagOut, err := exec.Output(exec.Command("gcloud", "compute", "instances", "list",
+			"--project="+project,
+			"--filter=metadata.created-by:*"+instanceGroups[project][cluster][0].path,
+			"--limit=1",
+			"--format=get(tags.items)"))
+		if err != nil {
+			return fmt.Errorf("instances list failed: %s", execError(err))
+		}
+		tag := strings.TrimSpace(string(tagOut))
+		if tag == "" {
+			return fmt.Errorf("instances list returned no instances (or instance has no tags)")
+		}
+
+		if err := runWithOutput(exec.Command("gcloud", "compute", "firewall-rules", "create", firewall,
+			"--project="+project,
+			"--network="+network,
+			"--allow="+e2eAllow,
+			"--target-tags="+tag)); err != nil {
+			return fmt.Errorf("error creating e2e firewall rule: %v", err)
+		}
 	}
 	return nil
 }
 
-func (d *deployer) getClusterFirewall(project, cluster string) string {
+func getClusterFirewall(project, cluster string, instanceGroups map[string]map[string][]*ig) string {
 	// We want to ensure that there's an e2e-ports-* firewall rule
 	// that maps to the cluster nodes, but the target tag for the
 	// nodes can be slow to get. Use the hash from the lexically first
 	// node pool instead.
-	return "e2e-ports-" + d.instanceGroups[project][cluster][0].uniq
+	return "e2e-ports-" + instanceGroups[project][cluster][0].uniq
+}
+
+// Ensure firewall rules for multi-project profile.
+// Reference: https://cloud.google.com/kubernetes-engine/docs/how-to/cluster-shared-vpc#creating_additional_firewall_rules
+// Please note we are not including the firewall rule for SSH connection as it's not needed for testing.
+func ensureFirewallRulesForMultiProjects(projects []string, network string, subnetworkRanges []string) error {
+	hostProject := projects[0]
+	for i := 1; i < len(projects); i++ {
+		curtProject := projects[i]
+		firewall := fmt.Sprintf("%s-rule-%s", hostProject, curtProject)
+		// sourceRanges need to be separated with ",", while the provided subnetworkRanges are separated with space.
+		sourceRanges := strings.ReplaceAll(subnetworkRanges[i-1], " ", ",")
```

Comment:
do we need them space separated as input?
let's just get them comma separated as input too?

(we should still keep this logic for backwards compatibility)

Reply:
We are using comma to separate the subnetwork ranges for different projects, see https://github.com/kubernetes-sigs/kubetest2/blob/master/kubetest2-gke/deployer/deployer.go#L252, so we cannot change each range to be comma separated.